Mathlib/Algebra/Lie/EngelSubalgebra.lean

```diff
@@ -0,0 +1,158 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import Mathlib.Algebra.Lie.Engel
+import Mathlib.Algebra.Lie.Normalizer
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.Algebra.Lie.Subalgebra
+import Mathlib.Algebra.Lie.Weights.Basic
+import Mathlib.Data.Finset.NatAntidiagonal
+
+/-!
+# Engel subalgebras
+
+This file defines Engel subalgebras of a Lie algebra and provides basic related properties.
+
+The Engel subalgebra `LieSubalgebra.Engel R x` consists of
+all `y : L` such that `(ad R L x)^n` kills `y` for some `n`.
+
+## Main results
+
+Engel subalgebras are self-normalizing (`LieSubalgebra.normalizer_engel`),
+and minimal ones are nilpotent (TODO), hence Cartan subalgebras.
+
+* `LieSubalgebra.normalizer_eq_self_of_engel_le`:
+  Lie subalgebras containing an Engel subalgebra are self-normalizing,
+  provided the ambient Lie algebra is artinian.
+* `LieSubalgebra.isNilpotent_of_forall_le_engel`:
+  A Lie subalgebra of a noetherian Lie algebra is nilpotent
+  if it is contained in the Engel subalgebra of all its elements.
+-/
+
+open BigOperators LieAlgebra LieModule
+
+variable {R L M : Type*} [CommRing R] [LieRing L] [LieAlgebra R L]
+  [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+
+namespace LieSubalgebra
+
+variable (R)
+
+/-- The Engel subalgebra `Engel R x` consists of
+all `y : L` such that `(ad R L x)^n` kills `y` for some `n`.
+
+Engel subalgebras are self-normalizing (`LieSubalgebra.normalizer_engel`),
+and minimal ones are nilpotent, hence Cartan subalgebras. -/
+@[simps!]
+def Engel (x : L) : LieSubalgebra R L :=
```

Comment:
I _think_ initial uppercase is only for `Prop`-valued definitions.
```suggestion
def engel (x : L) : LieSubalgebra R L :=
```

Reply:
Thanks, fixed everywhere.